File: CppGen/CppGen/ExternalFunction.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace CppGen
{
	// Stores an external function whenever external_define is called.
	class ExternalFunction
	{
		public string Name;
		public DataType ReturnType;
		public List<DataType> ArgTypes;

		public ExternalFunction(string name, DataType returnType, List<DataType> argTypes)
		{
			Name = name;
			ReturnType = returnType;
			ArgTypes = argTypes;
		}

		// Writes the C++ header of the external function.
		public void WriteCppHeader()
		{
			CodeWriter.Write(ReturnType.ToCpp() + " " + CodeObject.NameToCpp(Name));

			CodeWriter.Write("(");
			int a = 0;
			foreach (DataType type in ArgTypes)
			{
				if (a++ > 0)
					CodeWriter.Write(", ");
				CodeWriter.Write(type.ToCpp());
			}
			CodeWriter.WriteLine(");");
		}
	}
}

File: CppGen/Sources/ExternalFunction.cpp
#include "CppGen.hpp"

namespace CppGen
{
ExternalFunction::ExternalFunction(StringId name, DataType* returnType, List<DataType*> argTypes)
{
	this->name = name;
	this->returnType = returnType;
	this->argTypes = argTypes;
}

void ExternalFunction::writeCppHeader()
{
	CodeWriter::write(this->returnType->toCpp() + " " + CodeObject::nameToCpp(this->name));

	CodeWriter::write("(");
	int a = 0;
	for (DataType* type : this->argTypes)
	{
		if (a++ > 0)
			CodeWriter::write(", ");
		CodeWriter::write(type->toCpp());
	}
	CodeWriter::writeLine(");");
}

}
